src/animation/KeyframeTrack.js

```diff
@@ -106,79 +182,109 @@ THREE.KeyframeTrack.prototype = {
 
 	// removes keyframes before and after animation without changing any values within the range [startTime, endTime].
 	// IMPORTANT: We do not shift around keys to the start of the track time, because for interpolated keys this will change their values
- 	trim: function( startTime, endTime ) {
+	trim: function( startTime, endTime ) {
+
+		var times = this.times;
+		var nKeys = times.length;
 
 		var firstKeysToRemove = 0;
-		for ( var i = 1; i < this.keys.length; i ++ ) {
-			if ( this.keys[i] <= startTime ) {
-				firstKeysToRemove ++;
-			}
+		for ( var i = 1; i !== nKeys; ++ i ) {
+
+			if ( times[i] <= startTime ) ++ firstKeysToRemove;
+
 		}
 
 		var lastKeysToRemove = 0;
-		for ( var i = this.keys.length - 2; i > 0; i ++ ) {
-			if ( this.keys[i] >= endTime ) {
-				lastKeysToRemove ++;
-			} else {
-				break;
-			}
+		for ( var i = nKeys - 2; i !== 0; -- i ) {
+
+			if ( times[i] >= endTime ) ++ lastKeysToRemove;
+			else break;
+
 		}
 
 		// remove last keys first because it doesn't affect the position of the first keys (the otherway around doesn't work as easily)
-		if ( ( firstKeysToRemove + lastKeysToRemove ) > 0 ) {
-			this.keys = this.keys.splice( firstKeysToRemove, this.keys.length - lastKeysToRemove - firstKeysToRemove );;
-		}
+		if( ( firstKeysToRemove + lastKeysToRemove ) !== 0 ) {
 
-		return this;
+			var from = firstKeysToRemove;
+			var to = nKeys - lastKeysToRemove - firstKeysToRemove;
 
-	},
+			this.times = THREE.AnimationUtils.arraySlice( times, from, to );
 
-	/* NOTE: This is commented out because we really shouldn't have to handle unsorted key lists
-	         Tracks with out of order keys should be considered to be invalid.  - bhouston
-	sort: function() {
+			var values = this.values;
+			var stride = this.getValueSize();
+			this.values = THREE.AnimationUtils.arraySlice( values, from * stride, to * stride );
 
-		this.keys.sort( THREE.KeyframeTrack.keyComparer );
+		}
 
 		return this;
 
-	},*/
+	},
 
 	// ensure we do not get a GarbageInGarbageOut situation, make sure tracks are at least minimally viable
 	// One could eventually ensure that all key.values in a track are all of the same type (otherwise interpolation makes no sense.)
 	validate: function() {
 
-		var prevKey = null;
+		if ( this.getValueSize() % 1 !== 0 ) {
```

Comment:
> See the extensibility note in BooleanKeyframeTrack for an example use case.

There is actually a huge amount of literature on compressed animation formats.  Compressing the boolean tracks that way is nice, but it should be an option because they are not easily editable after that is done.  I'd suggest having new classes to represent the compressed formats if their internal formats differ.

The optimize function for keyframe tracks can easily be extended to do much more useful optimizations, but again this isn't that useful in the context of playback because you really do not want to do it at load time.

In the past I've written linear optimizers with thresholds -- it just iteratively looks at keys that can be removed that can be reconstructed by linear interpolations of the resulting keyframes.  There is more here:

https://docs.unrealengine.com/latest/INT/Engine/Animation/Sequences/index.html#compression

Some of the more advanced stuff applies compression on the data itself using size optimal representaions, rather than just removing keys:

https://engineering.riotgames.com/news/compressing-skeletal-animation-data

But all this should not be part of this PR.